### ISSUE TITLE
Non-deterministic behavior of BoundaryNormalLFIntegrator with a VectorGridFunctionCoefficient

### ISSUE THREAD

**Issue by salazardetroya** [mfem/mfem#1413]
I want to perform a boundary integral of a flux: `\int_{\Gamma} f \cdot n d\Gamma` where `\Gamma` is a boundary with a specific marker, `f` is a vector and `n` is the surface normal. I decided to use `BoundaryNormalLFIntegrator` for this purpose. In my problem the vector `f` will come from a `ParGridFunction`, hence the need to use `VectorGridFunctionCoefficient`, although in this example I am just defining my `ParGridFunction` with a `VectorConstantCoefficient` just to have a minimum code to reproduce the error.

The following code:
```cpp
#include "mfem.hpp"
using namespace mfem;
int main(int argc, char *argv[])
{
    int num_procs, myid;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myid);

    const char *mesh_file = "beam-quad.mesh";

    Mesh *mesh = new Mesh(mesh_file, 1, 1);
    int dim = mesh->Dimension();
    int ref_levels = 5;
    for (int l = 0; l < ref_levels; l++) {
        mesh->UniformRefinement();
    }

    ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
    delete mesh;

    auto h1_fec = std::make_shared<H1_FECollection>(1, dim);
    auto h1_fespace =
        std::make_shared<ParFiniteElementSpace>(pmesh, h1_fec.get());
    auto h1_fespace_flux =
        std::make_shared<ParFiniteElementSpace>(pmesh, h1_fec.get(), dim);

    Vector flux_vec(2);
    flux_vec(0) = 1.0;
    flux_vec(1) = 1.0;
    VectorConstantCoefficient flux_coeff(flux_vec);

    ParGridFunction flux_func(h1_fespace_flux.get());
    flux_func.ProjectCoefficient(flux_coeff);

    VectorGridFunctionCoefficient flux_coef_vec(&flux_func);

    Array<int> flux_bdr(pmesh->bdr_attributes.Max());
    flux_bdr = 0;
    flux_bdr[1] = 1;

    auto flux_lf = std::make_shared<ParLinearForm>(h1_fespace.get());
    flux_lf->AddBoundaryIntegrator(new BoundaryNormalLFIntegrator(flux_coef_vec), flux_bdr);
    flux_lf->Assemble();
    ParGridFunction one_gf(h1_fespace.get());
    ConstantCoefficient onecoeff(1.0);
    one_gf.ProjectCoefficient(onecoeff);
    double flux = flux_lf->operator()(one_gf);
    std::cout << "flux value: " << flux << std::endl;
    delete pmesh;
}
```

seems to behave erratically:
```bash
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 1
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 1
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 1
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 1
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 1
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 0.5
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: nan
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 1
salazardetro1@monticello:~/myprojects/TransportER/amr-study/build$ ./test_boundary_integral
flux value: 0.5
```

**Comment by mlstowell** [comment]
Hi, @salazardetroya ,
I'm not able to reproduce the problem.  I tried your test program several times in both serial and parallel but it always returned an answer of 1.

Whenever I run into issues like this I always try to rebuild MFEM (or other libraries that may have changed) starting from `make clean`.  Often the root of the problem is a small change to a header file which alters the data layout in memory so that outdated object files access the wrong data.  Build systems don't always properly determine dependencies on header files.

Best wishes,
Mark

**Comment by salazardetroya** [comment]
I cleaned up everything and compiled `master` from scratch. I still get the same random behavior. I compiled mfem with no MPI and it works. I am going to try to see if different MPI implementations work (I was using openmpi-3.1.4 downloaded by petsc-3.12.4)

**Comment by salazardetroya** [comment]
Still no luck. I have left it aside and started working using the LC machines, where it works well. @mlstowell I think to remember you have MFEM on your Mac, correct? Which MPI implementation do you have and how you installed it? Same for compilers if different than the standard.

**Comment by mlstowell** [comment]
Hi, @salazardetroya ,
I'm sorry to hear that.  I do run MFEM on a Mac. Currently, I'm using mpich version 3.3.2 and gcc version 9.3.0 both installed through Homebrew.  The Homebrew version is 2.2.3 which I installed through the lab's MacPatch tool (select "COMP Scholar Program" from the pull-down menu).

Best wishes,
Mark

**Comment by jandrej** [comment]
Reproducable on

Apple LLVM version 10.0.1 (clang-1001.0.46.4)
OpenMPI 4.0.2 (homebrew install)
CMake build system

```
/Library/Developer/CommandLineTools/usr/bin/c++   -I/*/mfem-master/build-opt -I/*/open-mpi/4.0.2/include  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk   -std=gnu++11 -o CMakeFiles/bugtest.dir/bugtest.cpp.o -c bugtest.cpp
```
```
dyro% for i in {1..10000}; do ./bugtest; done
flux value: 1
flux value: 1
flux value: 1
flux value: nan
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: 1
flux value: nan
flux value: 1
flux value: 1
flux value: 1
flux value: nan
```

**Comment by psocratis** [comment]
I can reproduce too, even in a serial run

```cpp
#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;
int main(int argc, char *argv[])
{
    const char *mesh_file = "../data/beam-quad.mesh";

    Mesh *mesh = new Mesh(mesh_file, 1, 1);
    int dim = mesh->Dimension();
    int ref_levels = 5;
    for (int l = 0; l < ref_levels; l++) {
        mesh->UniformRefinement();
    }


    FiniteElementCollection *h1_fec = new H1_FECollection(1,dim);
    FiniteElementSpace *h1_fespace = new FiniteElementSpace(mesh, h1_fec);
    FiniteElementSpace *h1_fespace_flux = new FiniteElementSpace(mesh, h1_fec,dim);

    Vector flux_vec(2);
    flux_vec(0) = 1.0;
    flux_vec(1) = 1.0;
    VectorConstantCoefficient flux_coeff(flux_vec);

    GridFunction flux_func(h1_fespace_flux);
    flux_func.ProjectCoefficient(flux_coeff);

    VectorGridFunctionCoefficient flux_coef_vec(&flux_func);

    Array<int> flux_bdr;
    if (mesh->bdr_attributes.Size())
    {
        flux_bdr.SetSize(mesh->bdr_attributes.Max());
        flux_bdr = 0;
        if (mesh->bdr_attributes.Max()>1) flux_bdr[1] = 1;
    }

    LinearForm * flux_lf = new LinearForm(h1_fespace);
    flux_lf->AddBoundaryIntegrator(new BoundaryNormalLFIntegrator(flux_coef_vec), flux_bdr);
    flux_lf->Assemble();
    GridFunction one_gf(h1_fespace);
    ConstantCoefficient onecoeff(1.0);
    one_gf.ProjectCoefficient(onecoeff);

    double flux = flux_lf->operator()(one_gf);
    std::cout << "flux value: " << flux << std::endl;
    
    delete h1_fespace_flux;
    delete h1_fespace;
    delete h1_fec;
    delete mesh;
}
}
```

**Comment by psocratis** [comment]
seems like something is causing a problem with the integration points
when it fails I get integration points like x = 0.211325, y = -3.10504e+231, z = 2.44141e+232 instead of x = 0.211325, y = 0, z = 0

**Comment by psocratis** [comment]
So it seems that the method IntegrationPoint::Init is not called properly, and that causes the random numbers in y and z values of the integration point. I changed this in intrules.hpp

```cpp
class IntegrationPoint
{
public:
   double x, y, z, weight;
   int index;

   void Init(int const i)
   {
      x = y = z = weight = 0.0;
      index = i;
   }
````

to this 

```cpp
class IntegrationPoint
{
public:
   double x=0.0;
   double y=0.0;
   double z=0.0;
   double w=0.0;
   double weight=0.0;

   int index;

   void Init(int const i)
   {
      x = y = z = weight = 0.0;
      index = i;
   }
````
and the issue goes away but still this is not the best thing to do cause the index is not set. Can anyone who is more familiar with this class help?

**Comment by salazardetroya** [comment]
Shouldn't have these unitializations been caught by the static analyzer?

**Comment by mlstowell** [comment]
Also, I would expect the integration point to be set by the integrator, which would overwrite the initial values.

**Comment by mlstowell** [comment]
@psocratis , this is using a 1D integration rule so only `ip.x` is being set.  It's possible that `ip.y` and/or `ip.z` are being accessed but this would probably indicate an error.

**Comment by psocratis** [comment]
> @psocratis , this is using a 1D integration rule so only `ip.x` is being set. It's possible that `ip.y` and/or `ip.z` are being accessed but this would probably indicate an error.

You are right, that's why this is a bit puzzling. More specifically this function is called.

````cpp
void QuadratureFunctions1D::GaussLegendre(const int np, IntegrationRule* ir)
{
   ir->SetSize(np);

   switch (np)
   {
      case 1:
         ir->IntPoint(0).Set1w(0.5, 1.0);
         return;
      case 2:
         ir->IntPoint(0).Set1w(0.21132486540518711775, 0.5);
         ir->IntPoint(1).Set1w(0.78867513459481288225, 0.5);
         return;
      case 3:
         ir->IntPoint(0).Set1w(0.11270166537925831148, 5./18.);
         ir->IntPoint(1).Set1w(0.5, 4./9.);
         ir->IntPoint(2).Set1w(0.88729833462074168852, 5./18.);
         return;
   }
````
if I initialize at case 2 the problem goes away. Which means y coordinate is used.

**Comment by mlstowell** [comment]
Why would case 2 indicate that the y coordinate is used?

**Comment by psocratis** [comment]
````cpp
void BoundaryNormalLFIntegrator::AssembleRHSElementVect(
   const FiniteElement &el, ElementTransformation &Tr, Vector &elvect)
{
   int dim = el.GetDim()+1;
   int dof = el.GetDof();
   Vector nor(dim), Qvec;

   shape.SetSize(dof);
   elvect.SetSize(dof);
   elvect = 0.0;

   const IntegrationRule *ir = IntRule;
   if (ir == NULL)
   {
      int intorder = oa * el.GetOrder() + ob;  // <----------
      ir = &IntRules.Get(el.GetGeomType(), intorder);
   }

   for (int i = 0; i < ir->GetNPoints(); i++)
   {
      const IntegrationPoint &ip = ir->IntPoint(i);

      Tr.SetIntPoint(&ip);
      CalcOrtho(Tr.Jacobian(), nor);
      Q.Eval(Qvec, Tr, ip);

      el.CalcShape(ip, shape);

      elvect.Add(ip.weight*(Qvec*nor), shape);
   }
}
````




> Why would case 2 indicate that the y coordinate is used?

**Comment by psocratis** [comment]
> Why would case 2 indicate that the y coordinate is used?

In this particular run the code goes in case 2. Basically setting two integration points. 
If before setting it you initialize y and z then the problem goes away.

**Comment by psocratis** [comment]
Additionally if you look at the integrator 
````cpp
void BoundaryNormalLFIntegrator::AssembleRHSElementVect(
   const FiniteElement &el, ElementTransformation &Tr, Vector &elvect)
{
   int dim = el.GetDim()+1;
   int dof = el.GetDof();
   Vector nor(dim), Qvec;

   shape.SetSize(dof);
   elvect.SetSize(dof);
   elvect = 0.0;

   const IntegrationRule *ir = IntRule;
   if (ir == NULL)
   {
      int intorder = oa * el.GetOrder() + ob;  // <----------
      ir = &IntRules.Get(el.GetGeomType(), intorder);
   }

   for (int i = 0; i < ir->GetNPoints(); i++)
   {
      const IntegrationPoint &ip = ir->IntPoint(i);

      Tr.SetIntPoint(&ip);
      CalcOrtho(Tr.Jacobian(), nor);
      Q.Eval(Qvec, Tr, ip);

      el.CalcShape(ip, shape);

      elvect.Add(ip.weight*(Qvec*nor), shape);
   }
}

Q is a vectorcoefficient and its evaluation involves the y coordinate. Only after the evaluation it's doted with the nor. 
````

**Comment by psocratis** [comment]
Hi @salazardetroya , It seems that this behavior is caused by the GridFunctionCoefficient and its evaluation on the boundary. This issue is addressed (and fixed) in #1194. 

Please let us know if you still encounter the problem.

**Comment by salazardetroya** [comment]
Thanks! I'll use that branch.
